File: packages/engine/src/cpus/6502/instructions/php.js
import { CycleType } from '..'

/**
 * PusH Processor flags (in the stack)
 *
 * @example
 * MODE          SYNTAX        HEX LEN TIM
 * Implied       PHP           $08  1   3
 */
export function* php() {
  this.fetch()
  if (--this.pc < 0) this.pc = 0xffff
  yield CycleType.GET

  this.write(0x100 | this.sp, this.status | 0x10)
  if (--this.sp < 0) this.sp = 0xff
  yield CycleType.PUT
}

File: packages/engine/src/cpus/6502/instructions/php.zig
const std = @import("std");
const CPU = @import("../cpu.zig").CPU;
const Addressing = @import("../opcode.zig").Addressing;

// PusH Processor flags (in the stack)
//
// MODE          SYNTAX        HEX LEN TIM
// Implied       PHP           $08  1   3

pub fn php(self: *CPU, addressing: Addressing) void {
    _ = addressing;
    if (self.cycle_counter == 2) {
        _ = self.fetch();
        self.pc -%= 1;
        self.next_cycle = .write;
        return;
    } else {
        self.write(0x100 | @as(u16, self.sp), self.getP() | 0x10);
        self.sp -%= 1;
    }
}
